Size how much printable data gets printed by Raw

yadij opened the pull request:
An arbitrary default of 256 bytes has been chosen. Callers can use the
whole() method to dump the entire data if necessary.

We size the number of characters to print rather than the total
data size of the output. Encodings (e.g., hex()) and decorations (e.g. labels and
truncation indicators) may result in more output characters than the
data size. For example, a 100-octet buffer sized to 10 octets would be
printed using 20 hex printables.

yadij committed:
Add limit parameter to Raw debug class

To restrict how much of the raw output gets printed when size_ is very large.

rousskov commented on src/Debug.h:
```diff
@@ -184,8 +184,8 @@ void _db_rotate_log(void);
 public:
-    Raw(const char *label, const char *data, const size_t size):
-        level(-1), label_(label), data_(data), size_(size), useHex_(false), useGap_(true) {}
+    Raw(const char *label, const char *data, const size_t size, const size_t limit = 0):
```
Please remove the limit from the constructor arguments. Raw uses dedicated methods for configuration options like that (e.g., hex(), gap(), and minLevel()). Please add such a method if at least one caller needs it. I suggest calling it atMost(), avoiding an overly generic "limit".

yadij replied:
done in dc3814d

rousskov commented on src/debug.cc:
```diff
@@ -876,7 +876,8 @@ Raw::printHex(std::ostream &os) const
     const auto savedFlags = os.flags(); // std::ios_base::fmtflags
     os << std::hex;
-    std::for_each(data_, data_ + size_,
+    auto len = (limit_ > 0 && limit_ < size_ ? limit_ : size_);
```
Please add a printableSize() or similar method to avoid duplicating this code.

yadij replied:
```suggestion
```

rousskov commented on src/Debug.h:
```diff
@@ -214,6 +214,7 @@ class Raw
     const char *label_; ///< optional data name or ID; triggers size printing
     const char *data_; ///< raw data to be printed
     size_t size_; ///< data length
+    size_t limit_; ///< do not print more than this amount of the data
```
```suggestion
    size_t limit_; ///< do not print more data (0 means unlimited)
```

rousskov commented on src/debug.cc:
```diff
@@ -902,8 +903,10 @@ Raw::print(std::ostream &os) const
-                os.write(data_, size_);
+            else {
+                auto len = (limit_ > 0 && limit_ < size_ ? limit_ : size_);
+                os.write(data_, len);
```
```suggestion
                os.write(data_, printableSize());
```

rousskov commented on src/debug.cc:
```diff
@@ -876,7 +876,8 @@ Raw::printHex(std::ostream &os) const
     os << std::hex;
-    std::for_each(data_, data_ + size_,
+    auto len = (limit_ > 0 && limit_ < size_ ? limit_ : size_);
+    std::for_each(data_, data_ + len,
```
```suggestion
    std::for_each(data_, data_ + printableSize(),
```

rousskov commented on src/debug.cc:
```diff
@@ -902,8 +903,10 @@ Raw::print(std::ostream &os) const
+            else {
+                auto len = (limit_ > 0 && limit_ < size_ ? limit_ : size_);
+                os.write(data_, len);
+            }
```
Please provide a visual clue to the reader that the data was truncated:

```suggestion
        if (printableSize() < size_)
            os << "[...]";
```

You can even print the number of truncated bytes: `[... N more bytes ...]`.

yadij replied:
Done in 33cdf16

rousskov reviewed (changes requested):
> A quick check indicates most current callers will not find a limit useful - so the initial default is no limiting.

I believe we should emphasize/prioritize logging safety/sanity _and_ flag large printers by specifying a large but still reasonable limit as the default (instead of no limit). There should be very few cases where printing more than, say, 1000 bytes of data is desirable/intentional. The default should reflect _that_ principle (rather than the fact that most Raw users either print small strings or deliberately print potentially large strings). Selecting a reasonable default limit now may also help minimize the time wasted on arguing about specific limits applied to individual borderline callers.

Please add a whole() method to be used by "large" printers to remove the default limit discussed above.

yadij commented on src/debug.cc:
```diff
@@ -913,8 +914,10 @@ Raw::print(std::ostream &os) const
-            else
-                os.write(data_, size_);
+            else {
+                auto len = (limit_ > 0 && limit_ < size_ ? limit_ : size_);
```
```suggestion
```

yadij commented on src/debug.cc:
```diff
@@ -913,8 +914,10 @@ Raw::print(std::ostream &os) const
+            else {
+                auto len = (limit_ > 0 && limit_ < size_ ? limit_ : size_);
+                os.write(data_, len);
+            }
```
```suggestion
```

yadij commented on src/debug.cc:
```diff
@@ -913,8 +914,10 @@ Raw::print(std::ostream &os) const
                 printHex(os);
-            else
-                os.write(data_, size_);
+            else {
```
```suggestion
            else
```

rousskov commented on src/Debug.h:
```diff
@@ -214,10 +217,12 @@ class Raw
 
 private:
     void printHex(std::ostream &os) const;
+    size_t printableSize() const { return (limit_ > 0 && limit_ < size_ ? limit_ : size_); }
```
AFAICT, after the other changes, this method should be removed while limit_ is renamed to printableSize_. It is cheaper and clearer to compute this value correctly once than to call this method two-three times in a row. This change will also clarify what `limit_` is to the code reader (without the need to consult the member declaration).

BTW, the atMost() method name should stay the same. From the caller point of view, that method sets a limit which may (and often will!) exceed `size_`. The _implementation_ of that method should set `printableSize_` correctly, of course.

yadij replied:
Done in f06e16fe5

rousskov reviewed (changes requested):
The primary API problem identified in my previous [review](https://github.com/squid-cache/squid/pull/466#pullrequestreview-284241071) still needs to be addressed. The rest is polishing, including another API addition request from the previous review. I tried to itemize all these changes in the inlined requests now. Please let me know if you want me to finish all these changes instead.

> TODO full review of all current uses of class Raw to see which can be limited for better logs

I think this review should be done during this PR. We may decide not to change any other callers, of course, but I see no reason to postpone this review, especially since there are probably fewer than 30 occurrences to review.

The word "full" may imply some complex analysis. I do not think anything like that is needed right now. The focus of this review should be on cases that may have printed more than the default print size -- should they call whole() now? Or specify some larger-than-the-default limit?

Besides making this PR work complete, the review of current Raw users may inspire a different default limit than the currently proposed semi-arbitrary 1000 bytes.

rousskov commented:
> Alex: We may decide not to change any other callers, of course

The "other" part was wrong -- this PR currently changes no callers. However, there is at least one caller that should be changed -- HttpHeaderEntry::parse(). The need for a printing limit in that caller [triggered](https://github.com/squid-cache/squid/pull/445#discussion_r316909763) this PR! Completing that TODO item may identify other callers, of course.

rousskov commented:
@yadij, I was surprised to see this PR closed. Limiting Raw output size is clearly a useful feature, needed in existing and future code. Both of us spent quite a bit of time on this PR. Can you detail why you decided to close this PR, please?

yadij commented:
Small 'oops' cleaning up my local repo. It seems to have removed branches and closed PRs which I only had a copy on my company testing rig. Trying to fix these now.

yadij commented on src/debug.cc:
```diff
@@ -914,7 +914,9 @@ Raw::print(std::ostream &os) const
-                os.write(data_, size_);
+                os.write(data_, printableSize());
+            if (printableSize() < size_)
+                os << "[..." << (size_ - printableSize()) << " bytes...]";
```
"more" is implied by the '...', so not adding this extra word.

yadij commented on src/Debug.h:
```diff
@@ -189,11 +189,14 @@ class Raw
     Raw &minLevel(const int aLevel) { level = aLevel; return *this; }
 
+    /// print no more than n bytes of data (0 means unlimited)
+    Raw &atMost(const size_t n) { limit_ = n; return *this; }
```
Done in 6a29f9520

yadij commented on src/Debug.h:
```diff
@@ -189,11 +189,14 @@ class Raw
 public:
     Raw(const char *label, const char *data, const size_t size):
-        level(-1), label_(label), data_(data), size_(size), useHex_(false), useGap_(true) {}
+        level(-1), label_(label), data_(data), size_(size), limit_(0), useHex_(false), useGap_(true) {}
```
Done in 6a29f9520

rousskov commented:
I have updated PR title/description to better match the current PR code. Primary changes:

* Updated example to match the current code.

* Removed stale TODO: Since the new functionality is now enabled by default, the "full review" of all current uses to add limits is no longer necessary. We should still review current uses during this PR, but that work does not need a TODO in the PR description.

* Re-formatted.

Please adjust further as needed. I will post a review shortly.

rousskov reviewed (changes requested):
I think we are getting close to being done here. These remaining Raw changes are just follow-up adjustments after the default-to-limit and zero-is-not-special changes in the earlier reviews. I am happy to do this round myself if you prefer; just let me know.

Related Raw-using code fixes that I know about are:

* Downloader::CbDialer::print()
* "ignoring huge header field" in HttpHeaderEntry::parse() that [triggered](https://github.com/squid-cache/squid/pull/445#discussion_r316909763) this PR
* "found header with invalid characters" in HttpHeaderEntry::parse()
* `REQUEST:` and `RESPONSE:` whole-header dumping code; the ones using Raw will need `.whole()` adjustments, right?

Since we seem to have settled on the API, now is probably a good time to add those changes to the PR. I can do these as well if you prefer; just let me know.

yadij committed:
Update src/base/Raw.h

Co-authored-by: Alex Rousskov <rousskov@measurement-factory.com>

yadij committed:
Update Raw.h

yadij committed:
Update Raw.cc

yadij committed:
Update callers of Raw()

Audited all of the current callers. These ones looked like
they needed a different limit than the new default 256 bytes.

yadij commented:
> I think we are getting close to being done here. These remaining Raw changes are just follow-up adjustments after the default-to-limit and zero-is-not-special changes in the earlier reviews. I

In [2d45af8](https://github.com/squid-cache/squid/pull/466/commits/2d45af89f781daaa76f6877e1b283c505f679cd9) I have gone through all the calls to `Raw("` and adjusted the ones I think need to use a different limit from the default 256 being added here.   I also dropped a few `debugs()` lines in FTP where the buffer content would be printed twice in a row at different levels.

Please check if I missed any you think need changing, or disagree on keeping any of the old 100-byte limits.

kinkie reviewed (dismissed):
lgtm

yadij committed:
Implement TODO about Raw() constructor

rousskov commented on src/base/Raw.cc:
```diff
@@ -23,18 +23,22 @@ Raw::print(std::ostream &os) const
         return os;
 
     // finalize debugging level if no level was set explicitly via minLevel()
+    const auto printLimit = std::min(size_, printableSize_);
```
We never print fewer (or more) bytes than this variable, so this variable is not really a "limit" but "size":


```suggestion
    const auto printableDataSize = std::min(size_, printableDataSizeLimit_);
```

rousskov replied:
I do not insist on any specific name -- feel free to propose something better while still addressing my concern. For example, `reportableDataSize` and `reportableDataSizeLimit_` could also work (but I think "printable" is better here).

> this algorithm is not limited to "printable" bytes

The word "printable" has several interpretations. IMHO, the proposed name matches one of them, and humans would understand (with high probability) which meaning is used _in this specific context_.

> This is the limit bounding output of the action

It is not a limit; it is the _exact_ size -- we never send fewer data bytes to the printing code; we send exactly that many data bytes.

> The object itself may be larger ...
> [the caller may lie about the object size]

Yes, but those facts are not relevant to this particular variable.

> `asHex` feature means that we cannot consider this to be the "printable" size/amount of output either.

Yes, and that is why we should not call this variable "outputSize" or something like that. Suggested "printableDataSize" does _not_ imply that formatted/decorated output is going to have the same size.

yadij replied:
So, using your definitions described here to re-write the title and description.

rousskov commented on src/HttpHeader.cc:
```diff
@@ -1584,8 +1583,7 @@ HttpHeaderEntry::parse(const char *field_start, const char *field_end, const htt
         if (!CharacterSet::TCHAR[*pos]) {
-            debugs(55, 2, "found header with invalid characters in " <<
-                   Raw("field-name", field_start, min(name_len,100)) << "...");
+            debugs(55, 2, "found invalid characters in " << Raw("field-name", field_start, name_len).atMost(100));
```
Since we are changing this line, I would report the character position as well. Something like this may work (untested):


```suggestion
            debugs(55, 2, "found invalid character at position " << (pos-field_start) << " in " << Raw("field-name", field_start, name_len).atMost(100));
```

I do not insist on this improvement.

rousskov commented on src/base/Raw.h:
```diff
@@ -42,16 +49,17 @@ class Raw
+    const void *data_ = nullptr; ///< raw data to be printed
+    size_t size_ = 0; ///< data length
+    size_t printableSize_ = 256; ///< do not print more data
+    bool useHex_ = false; ///< whether hex() has been called
```
Please address `ErrorPage::BuildErrorPrinter::print()` TODO that was waiting for this "Add support for prefix printing to Raw" PR.

rousskov reviewed (changes requested):
> Please check if I missed any you think need changing

Done. I left a few specific change requests (all tied to an unrelated `bool useHex_` line).

> or disagree on keeping any of the old 100-byte limits.

We should drop them IMO: They were as arbitrary as the PR-proposed 256-byte limit, they were not really context-specific, 256-byte limit works OK for those cases, and I see no good reason to keep two arbitrary limits here. However, I do not insist on dropping those two `atMost(100)` calls.

I also removed the no longer true "removed the private method printHex() unimplemented since 675b840" claim from PR description.

yadij committed:
Update src/base/Raw.h

Co-authored-by: Alex Rousskov <rousskov@measurement-factory.com>

yadij committed:
Remove unused include

yadij commented on src/HttpHeader.cc:
```diff
@@ -1539,8 +1539,7 @@ HttpHeaderEntry::parse(const char *field_start, const char *field_end, const htt
         /* String must be LESS THAN 64K and it adds a terminating NULL */
-        // TODO: update this to show proper name_len in Raw markup, but not print all that
-        debugs(55, 2, "ignoring huge header field (" << Raw("field_start", field_start, 100) << "...)");
+        debugs(55, 2, "ignoring huge " << Raw("field-name", field_start, name_len).atMost(100));
```
This is not the entire header field, just the "field-name" syntax token.

yadij committed:
Remove 100 byte limits

Use the new 256 byte default instead.
Boith aree arbitraray amounts of output.

yadij committed:
Redo reviewer requests list in rebase

yadij committed:
Reviewer requested update

yadij commented on src/base/Raw.h:
```diff
@@ -42,16 +49,17 @@ class Raw
+    const void *data_ = nullptr; ///< raw data to be printed
+    size_t size_ = 0; ///< data length
+    size_t printableSize_ = 256; ///< do not print more data
+    bool useHex_ = false; ///< whether hex() has been called
```
Agreed and done in 137fd293d7.

Also, I am wondering WTF the `in` local has type `SBuf * const in` instead of `const SBuf &` ?

yadij committed:
fix log output for header warnings

yadij committed:
Reviewer requested change

yadij commented on src/base/Raw.h:
```diff
@@ -42,16 +49,17 @@ class Raw
+    const void *data_ = nullptr; ///< raw data to be printed
+    size_t size_ = 0; ///< data length
+    size_t printableSize_ = 256; ///< do not print more data
+    bool useHex_ = false; ///< whether hex() has been called
```
done in 4778dcd592

yadij committed:
Resolve TODO waiting on this feature

yadij committed:
Remove getStringPrefix()

down that slippery slope now that reviewer requires
old code that can use Raw() to do so.

yadij committed:
Reviewer request

yadij committed:
Fix build issues

squid-anubis commented:
Cannot create a git commit message from PR title and description.

Error while parsing PR description body: `the line is too long 81>72`

Problematic parser input:

    data size of the output. Encodings (e.g., hex()) and decorations (e.g. labels and
Please see PR [title](https://github.com/measurement-factory/anubis#pr-title) and [description](https://github.com/measurement-factory/anubis#pr-description) formatting requirements for more details.

This message was added by Anubis bot. Anubis will add a new message if the error text changes. Anubis will remove M-failed-description label when there are no corresponding failures to report.
